Add Page on Introduction to C#

YuPeiHenry opened the pull request:
Fixes #92

YuPeiHenry committed:
Add skeleton for C# and some background information.

YuPeiHenry committed:
Update C# features.

YuPeiHenry committed:
Add starting resources.

YuPeiHenry committed:
Fix errors.

YuPeiHenry commented:
@ChesterSng Can help me review? Thanks!

YuPeiHenry commented:
@sijie123 Can help me review from the perspective of someone new to C#? Thanks.

damithc commented:
Use `Fixes #...` in the PR description so that GitHub can auto-close the issue when the PR is merged.

sijie123 commented:
Hi Henry,
Thanks for the great guide to C#. I'll review this guide from the perspective of someone who doesn't know anything about C#, but has some experience in other OOP languages such as Java.

I think the background section is pretty detailed and useful. From the perspective of someone who doesn't know anything about C#, I would think that this is somewhat convincing for me to learn C#. 

I was taken a little by surprise when the next section was already the "noteworthy to C#" section. I thought it was a bit abrupt, as I was actually hoping for some more basic content / summaries / details on things within C#, especially more on its syntax. One reason why I haven't touched C# till now is because of its rather unique syntax. In particular, C# has things like namespaces, the use of colon for inheritance(?), square brackets, pound symbol and diamond notations that differ from what I understand in Java. The `public class A: B where B: C` syntax, for example, is something unique to C#, and might not be easy for newcomers to the language to understand what is going on.
The saving grace would be the resource at the bottom. I saw this resource to learning C#. Maybe we should get readers to have a prerequisite knowledge on these things before they read your guide.

That said, the section itself was really well written. Having used these features in Java before, (e.g. closure, optionals etc), the section was easy to read, understand and to apply.

Overall, good job in writing a detailed chapter, and keep it going! :)

damithc commented:
> I was taken a little by surprise when the next section was already the "noteworthy to C#" section. I thought it was a bit abrupt, as I was actually hoping for some more basic content / summaries / details on things within C#, especially more on its syntax.

This is a good point by @sijie123 
@YuPeiHenry Perhaps we can add a section to give a flavor of the C# language to the reader, a sort of a HelloWorld. But remember to write it in a way that you are describing the language, not teaching how to write C# code.

Also noticed that you are using 8 spaces for indentation? Use 4 instead, unless that's a C# standard.

ChesterSng commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,148 @@
+
+From Wikipedia:
+
+C# (pronounced C sharp) is a general-purpose, multi-paradigm programming language encompassing strong typing, lexically scoped, imperative, declarative, functional, generic, object-oriented (class-based), and component-oriented programming disciplines.
```
I am not unsure if the entire definition is needed, maybe you can phrase it your own words such that it easier for the audience to read and understand. 

Maybe you can combine it with 
> C# is an Object-Oriented Programming Language on the .NET framework, and uses Garbage Collection for memory management.

because they have some overlapping information too.

YuPeiHenry committed:
Update indentation.

ChesterSng commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,148 @@
+
+### Cross platform support
+
+C# is built on the .NET framework and has cross platform support extended by the (Mono project)[https://www.mono-project.com/], and similarly the complete runtime implementation from open source (CoreCL)[https://github.com/dotnet/coreclr].
```
Your brackets are wrong for the links

ChesterSng replied:
Maybe you can elaborate what does this mean for the user of C#

ChesterSng commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,148 @@
+
+## Why Learn C#
+
+C# is a relatively high level language and is not difficult to pick up (a little more difficult than Python and a little easier than Java.)
```
Maybe you can put this part under **Easy to Learn** subsection, and elaborate more on why it is not difficult

ChesterSng commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,148 @@
+
+From (Armina Mkhitaryan on Medium)[https://medium.com/sololearn/why-is-c-among-the-most-popular-programming-languages-in-the-world-ccf26824ffcb]
+
+Being powerful, flexible, and well-supported has meant C# has quickly become one of the most popular programming languages available.
```
Being powerful, flexible, and well-supported ~~has meant~~ enabled C# to quickly become one of the most popular programming languages available.

ChesterSng commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,148 @@
+//Get amount of fuel left
+Double fuelLeft = car?.GetFuel() ?? 0;
+//Explanation:
+//car?.GetFuel() evaluates to null when car is null.
```
maybe you can split this explanation to if  `car?` is null and if `car?` is not null

ChesterSng commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,148 @@
+
+### Default Interface Implementations
+
+Sometimes classes have a common interface but do not have a common ancestor and require the same method implmentation.
```
~~implmentation~~ implementation

ChesterSng commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,148 @@
+
+### Default Interface Implementations
+
+Sometimes classes have a common interface but do not have a common ancestor and require the same method implmentation.
```
I find this sentence a bit confusing, can a class have a common interface but not have a common ancestor?

YuPeiHenry replied:
If a class implements an interface without subclassing other classes. Unless interfaces are considered ancestors as well.

ChesterSng commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,148 @@
+A default implementation can be included in the interface to promote code reuse. Similarly, a certain class may not require certain methods listed in the interface.
+An example is a stub class that is required to implement an interface. A default method can be created with the `NotImplementedException`.
+
+The following links below describes in greater detail of some exciting features that can make your code more pleasant to read and write:
```
some exciting features, such as... (`async/await`, `delegates`...)
This will let the reader know what to expect in the links given!

YuPeiHenry committed:
Add flavor.

YuPeiHenry committed:
Shorten description.

YuPeiHenry committed:
Fix link formatting.

YuPeiHenry committed:
Fix links.

YuPeiHenry committed:
Fix typos.

YuPeiHenry committed:
Add clarity to various parts of article.

YuPeiHenry committed:
Change link text.

YuPeiHenry committed:
Add more comments to Glimpse of C#.

YuPeiHenry commented:
@sijie123 Thank you for the review. I have added a short section about how a simple C# program may look like, with an anchor link to the resource links at the end of the page. Let me know if it is still lacking.

YuPeiHenry commented:
@ChesterSng Thank you for your review. I have addressed the various formatting issues (typos, links, lack of context) in your review. There are 2 outstanding problems that I have yet to address:
* Easy to Learn subsection: I feel that the reasons for it being easy to learn is not sufficient. I have moved the sentence to `Glimpse of C#` section.
* interfaces: I'm not sure if a common interface is considered as a common ancestor, but interfaces used to not have method bodies (that was the point I was trying to get across).

ChesterSng commented:
@YuPeiHenry Hi Henry, there are a lot of jargons and hard-to-understand terms used in your book chapter, consider using some tooltips or modals to give a quick explanation of them!

sijie123 commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,173 @@
+    //Subclassing BaseClass
+    class MyClass : BaseClass {
+        ...
+		//Method declaration
```
The indentation looks off for this paragraph :O

YuPeiHenry replied:
Thanks for pointing this out.

sijie123 commented on contents/csharp/IntroductionToCSharp.md:
````diff
@@ -0,0 +1,173 @@
+
+Below is a sample code snippet of what a simple program in C# might look like:
+
+```
````
Is there a \`\`\`csharp or something similar to automatically syntax highlight the code?
I know you can do
\`\`\`java or \`\`\`js, but not sure about C#.

sijie123 commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,173 @@
+    * Parameterization and attributes
+
+Resources: [Introduction to Unit Testing](https://www.typemock.com/unit-test-patterns-for-net-part-i/)
+
```
What about XUnit?

sijie123 reviewed (changes requested):
Hi Henry! Since the last time I read this chapter, the "glimpse of C# syntax" has been very valuable and useful in giving me a nice leadup to C#. I think this strikes a good balance between "C# Tutorial" and "Intro to C#". Personally, I would have liked an explanation of for example why MyClass needs to extend BaseClass (<-- does every class need to eventually extend BaseClass? :O). That said, I don't think it's absolutely needed in this nice and short introduction, especially since it might move your chapter into more of a tutorial. So I think it's probably fine as it is.

As @ChesterSng mentioned, it would be good to sprinkle a couple of tooltips around jargons, if we want to assume that readers are completely new to OOP. I'm very familiar with these terms, so I can understand your chapter just fine. Perhaps add some tooltips around things like closure (<-- maybe an example of why it "may be useful to defer exection"?) or nullable (<-- what happens in conventional languages? how does optional make our lives easier?)

Other than the above, I think the chapter did a good job in convincing me why I should switch to C#, given its relevance and ease of use in many areas.
Keep up the good job :)

YuPeiHenry committed:
Add csharp syntax highlighting.

YuPeiHenry committed:
Correct indentation.

YuPeiHenry committed:
Add stub for XUnit.

YuPeiHenry committed:
Add tooltips.

YuPeiHenry committed:
Fix broken tooltip.

YuPeiHenry commented:
@sijie123 I have made changes to address the formatting issues. I have also added a short section on XUnit (haven't really used it before though). @ChesterSng I furnished most/all acronyms with tooltips).

YuPeiHenry committed:
Make minor ammendments.

YuPeiHenry commented:
Hi prof @damithc , can you help me do a preliminary review on the article?

YuPeiHenry committed:
Correct the title of the page.

YuPeiHenry committed:
Fix indentation.

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,186 @@
+Since it's such a robust and well-rounded language, it's no surprise that C# is utilized by thousands of companies. There are [5,000 C# jobs advertised in the US alone](https://gooroo.io/analytics/skill/C-Sharp#.WqipapPwYWo)
+(and 10,000 globally), with an [average base pay of nearly $80,000](https://www.glassdoor.com/Salaries/c-net-developer-salary-SRCH_KO0,15.htm).
+
+## Glimpse of C# Syntax
```
Remove heading and move the content under the heading `What is C#`

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,186 @@
+From [Armina Mkhitaryan on Medium](https://medium.com/sololearn/why-is-c-among-the-most-popular-programming-languages-in-the-world-ccf26824ffcb)
+
+Being powerful, flexible, and well-supported enabled C# has quickly become one of the most popular programming languages available.
+Today, it is the 4th most popular programming language, with approximately 31% of all developers using it regularly. It is also the 3rd largest community on StackOverflow (which was built using C#) with more than 1.1 million topics.
```
Use `<blockquote>` tag to quote text in the following format.

>Being powerful, flexible, and well-supported enabled C# has quickly become one of the most popular programming languages available.
>Today, it is the 4th most popular programming language, with approximately 31% of all developers using it regularly. It is also the 3rd largest community on StackOverflow (which was built using C#) with more than 1.1 million topics.
>--[_Why Is C# Among The Most Popular Programming Languages in The World?_ --Armina Mkhitaryan](https://medium.com/sololearn/why-is-c-among-the-most-popular-programming-languages-in-the-world-ccf26824ffcb)

damithc replied:
It's best to avoid giving numbers of current job count, ranking etc. as they get outdated quickly. When you give such numbers, give the date as well e.g., `(as at Jan 2019)`

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,186 @@
+# Introduction to C#
+
+**Author(s:) [Yu Pei, Henry](https://github.com/YuPeiHenry)**
+Reviewers: [Chester Sng](https://github.com/ChesterSng), [Lin Si Jie](https://github.com/sijie123)
```
There should be a line break between these two lines.

damithc commented on contents/csharp/IntroductionToCSharp.md:
````diff
@@ -0,0 +1,186 @@
+}
+```
+
+[Click here to get started](#how-to-get-started)
````
Probably not needed

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,186 @@
+* [Learn C#](https://www.sololearn.com/Course/CSharp/?ref=medcsharp) is a good resource to get started with C# and OOP.
+* [Common pitfalls](https://www.upwork.com/hiring/development/common-mistakes-in-c-sharp-programming/) covers some common mistakes made when programming in C#.
+* [Design Patterns in C#](https://www.c-sharpcorner.com/UploadFile/bd5be5/design-patterns-in-net/) introduces some common design patterns that can make planning and maintaining code easier.
+* [C# closures in greater depth](https://weblogs.asp.net/dixin/understanding-csharp-features-6-closure)
```
Consider this structure, to be consistent with other similar articles we are doing at the moment.

## What is C#?

Mention it is a _general purpose_, _multi-paradigm_, _garbage collected_, _cross-platform_ language by Microsoft, and part of the _.NET platform_.

Explain each of the characteristics mentioned above. e.g., one paragraph each. Can give a link to find more info about each but the article itself should contain a brief overview of each

Give the example code. Briefly mention how developers usually use VS to edit code and build it into an exe. you can also mean the level of tooling support e.g., built in testing support in VS

## C# Syntax Features
Mention that you are listing some noteworthy features that makes C# stand out.

### Closures

### Nullable types 

### etc.

## Why Learn C#

### Popular
Mention where C# is popular e.g., Windows desktop apps, Game development (due to Unity), anything else? 

### High in demand

## How to get started

Give a well thought out learning path.  Give commentary of each resource.

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,186 @@
+* https://stackify.com/csharp-8-features/
+* https://www.developer.com/net/csharp/slideshows/top-10-csharp-6.0-language-features.html
+
+### Testing facilities
```
I'm afraid this section doesn't fit into this article as it is odd to go into one aspect of development but not others. Besides, testing is not that unique to C#.  It fits better into an article about C# testing.

damithc commented on contents/csharp/IntroductionToCSharp.md:
````diff
@@ -0,0 +1,186 @@
+Foo foo = new Foo() {
+    MyBar = bar;
+};
+```
````
Can this be like this too (shorter)?
```csharp
Foo foo = new Foo() {
    MyBar = new Bar();
};
```

You can compare it with the longer version of the code to show the difference.
Also, as much as possible, use realistic examples rather than Foo/Bar.

damithc commented on contents/csharp/IntroductionToCSharp.md:
````diff
@@ -0,0 +1,186 @@
+
+Func<Int> counter = Context.GetCounter();
+counter();
+```
````
You need to explain this point in more details. The reader shouldn't have to go to external articles to be able to understand this article. External articles are for _additional_ info.

damithc commented on contents/csharp/IntroductionToCSharp.md:
````diff
@@ -0,0 +1,186 @@
+//If car is not null, one can expect the statement to evaluate to car.GetFuel().
+//If car is null, car?.GetFuel() evaluates to null
+//?? operator sees null, so the entire expression defaults to 0.
+```
````
This point too can benefit from more detailed explanations and code to compare _with_ and _without_ the feature.

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,186 @@
+The following links below describes in greater detail of some exciting features that can make your code more pleasant to read and write, such as `async/await`, `delegates` etc:
+* https://codeaddiction.net/articles/15/10-features-in-c-that-you-really-should-learn-and-use
+* https://stackify.com/csharp-8-features/
+* https://www.developer.com/net/csharp/slideshows/top-10-csharp-6.0-language-features.html
```
Similarly, this point need to be explained better, in the article itself (i.e., don't depend on the external links to explain the point).

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,186 @@
+
+C# is an [Object-Oriented Programming](https://en.wikipedia.org/wiki/Object-oriented_programming "Object-oriented programming (OOP) is a programming paradigm based on the concept of \"objects\", which can contain data, in the form of fields (often known as attributes), and code, in the form of procedures (often known as methods)")
+Language on the .NET framework, and uses [Garbage Collection](https://en.wikipedia.org/wiki/Garbage_collection_(computer_science) "In computer science, garbage collection (GC) is a form of automatic memory management. The garbage collector, or just collector, attempts to reclaim garbage, or memory occupied by objects that are no longer in use by the program.") for memory management.
+C# is considred a general-purpose, [multi-paradigm](https://en.wikipedia.org/wiki/Programming_paradigm "Programming paradigms are a way to classify programming languages based on their features. Languages can be classified into multiple paradigms.") programming language.
```
Perhaps mention that some claim C# to be Microsoft's answer to Java? Mentioning the similarity to Java can be useful to readers who are familiar with Java (our target audience is likely to be familiar with Java)

damithc reviewed (changes requested):
Good start. Content is mostly fine but needs better explanations in some places (instead of relying on external links to explain the point). Also, some restructuring of content can help.

YuPeiHenry committed:
Fix typo in closure's code block.

YuPeiHenry committed:
Resolve some simple formatting problems.

YuPeiHenry committed:
Add line spacing after author(s).

YuPeiHenry committed:
Replace blockquote with >

YuPeiHenry committed:
Fix spacing in blockquotes.

YuPeiHenry committed:
Remove -- from blockquote.

YuPeiHenry committed:
General restructuring of page.

YuPeiHenry committed:
Add relating statement from C# to Java.

YuPeiHenry committed:
Add dates to figures.

YuPeiHenry committed:
Add explanation for code block describing closures.

YuPeiHenry committed:
Add explanation and comparison to nullables code block.

YuPeiHenry committed:
Replace 8 spaces with 4 spaces.

YuPeiHenry committed:
Add explanation for async/await.

YuPeiHenry committed:
Expand on learning resources.

YuPeiHenry commented:
@damithc I have addressed the issues raised on the article. Can you give it another round of inspection? Thank you.

YuPeiHenry commented:
@sijie123 Do you mind helping me look through the article once more? I have reworked some parts of it to include more explanations.

YuPeiHenry committed:
Add minor formatting.

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,242 @@
+realisation. You may reduce these problems by reading about [some common mistakes in C# programming](https://www.upwork.com/hiring/development/common-mistakes-in-c-sharp-programming/).
+
+If you want to develop a desktop application for Windows, you may consult [Microsoft's documentation](https://docs.microsoft.com/en-us/visualstudio/get-started/csharp/tutorial-wpf?view=vs-2019)
+on creating an application with the `Windows Presentation Foundation`, a framework that is commonly used for creating UI for Windows Applications that has many useful features for building your UI.
```
Missing `</div>` at the end

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,242 @@
+
+## What is C#
+
+C# is a _general purpose_, _multi-paradigm_, _garbage collected_, _cross-platform_ language by Microsoft, and part of the _.NET platform_. Some claim C# is Microsoft's answer to Java.
```
> Some claim C# is Microsoft's answer to Java.

->

> Some claim C# is Microsoft's answer to Java due to the fact that the two languages have a lot of similarities.

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,242 @@
+## What is C#
+
+C# is a _general purpose_, _multi-paradigm_, _garbage collected_, _cross-platform_ language by Microsoft, and part of the _.NET platform_. Some claim C# is Microsoft's answer to Java.
+
```
Suggested to tweak as follows:

Given below are brief explanations of the key characteristics of C# mentioned above.
* **General purpose**: 
  >In computer software, a general-purpose programming language is a programming language designed to be used for writing software in the widest variety of application domains (a general-purpose language). A general-purpose programming language has this status because it does not include language constructs designed to be used within a specific application domain.<br>--(source: [Wikipedia](https://en.wikipedia.org/wiki/General-purpose_programming_language))
* **Multi-paradigm**: [_Programing Paradigms_](https://en.wikipedia.org/wiki/Programming_paradigm) are used to describe Programming Languages based on their features. Some commonly referred paradigms are [Object-Oriented Programming](https://en.wikipedia.org/wiki/Object-oriented_programming) (which primarily organizes code into objects that contain a state) and [Functional Programming](https://en.wikipedia.org/wiki/Functional_programming) (where code represents a sequence of stateless functions.)
C# supports both Object-Oriented and Functional Programming, and many others that can be found [here](https://en.wikipedia.org/wiki/C_Sharp_(programming_language)).
*

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,242 @@
+[Cross Platform](https://en.wikipedia.org/wiki/Cross-platform_software) software is software that can be run across multiple platforms, which may require recompilation depending on the software. Common platforms include Windows, MacOS and Linux, and for mobile platforms Android and iOS.
+The benefits of writing **Cross Platform** software is that developers will only need to primarily maintain 1 code base and be able to deploy to multiple platforms.
+
+### Glimpse of C#
```
`### A Glimpse of C#`?

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,242 @@
+
+### Glimpse of C#
+
+C# is a relatively high level language and is not difficult to pick up (a little more difficult than Python and a little easier than Java.)
```
`(a little more difficult than Python and a little easier than Java.)`: this is subjective. Best to avoid.

damithc replied:
We didn't mention _high-level_ in the list above. Can omit from here.

damithc commented on contents/csharp/IntroductionToCSharp.md:
````diff
@@ -0,0 +1,242 @@
+}
+```
+
+Developers that work with C# commonly use Visual Studio to maintain their code and also as a build tool for compilation into common application/library formats, such as `.exe` or `.dll`.
````
`to maintain their code`  -> `as their IDE`

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,242 @@
+
+## C# Syntax Features
+
+Most programming languages/frameworks have their own unique quirks, which can be utilised for better code quality. This section covers some **interesting and noteworthy features**
```
`quirks` is usually used to refer to negative things. Can do this:

>~~Most programming languages/frameworks have their own unique quirks, which can be utilised for better code quality.~~ This section covers some **interesting and noteworthy features** of C#.

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,242 @@
+## C# Syntax Features
+
+Most programming languages/frameworks have their own unique quirks, which can be utilised for better code quality. This section covers some **interesting and noteworthy features**
+that can help make code more concise, and most are common to other languages such as Java and Swift.
```
In MarkBind files we don't usually add line breaks to maintain a line within a certain length. i.e., there should not be a line break after `some **interesting and noteworthy features**`

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,242 @@
+Most programming languages/frameworks have their own unique quirks, which can be utilised for better code quality. This section covers some **interesting and noteworthy features**
+that can help make code more concise, and most are common to other languages such as Java and Swift.
+
+### Object / Array / Collection Initializers
```
`### Object/Array/Collection Initializers`

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,242 @@
+, [Array](https://www.webopedia.com/TERM/A/array.html "In programming, a series of objects all of which are the same size and type. Each object in an array is called an array element. For example,
+you could have an array of integers or an array of characters or an array of anything that has a defined data type.") 
+or [Collection](https://computersciencewiki.org/index.php/Collections "A collection — sometimes called a container — is simply an object that groups multiple elements into a single unit.") conveniently as shown.
+This can be useful when writing tests, as test data will be easier to read, as compared to calling the actual constructor.
```
You mention _convenient_ but doesn't explain why it is so. Can compare code with and without this feature?

damithc commented on contents/csharp/IntroductionToCSharp.md:
````diff
@@ -0,0 +1,242 @@
+counter(); //Returns 2
+```
+
+If you wish to read more about closures, you may consult [this article by dixin](https://weblogs.asp.net/dixin/understanding-csharp-features-6-closure)
````
The text does not explain why this section is named _Closures_

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,246 @@
+access, many if checks might be used. However, when dealing with nullable types, such code can be shortened
+while remaining easy to read, using collaescing operators `??`, or null conditional operators `.?`. Some may find this similar to `optionals` in Swift.
+
+More can be read about Nullables [here](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/nullable-types/)
```
The content above is premature. Follow this flow:
* how do we normally handle objects that can be null
* how does C# does it and why it is better
* where to find more info

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,246 @@
+//Explanation:
+//If car is not null, one can expect the statement to evaluate to car.GetFuel().
+//If car is null, car?.GetFuel() evaluates to null
+//?? operator sees null, so the entire expression defaults to 0.
```
This explanation can be in the normal text (i.e., not in code comments)

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,246 @@
+
+Sometimes classes have a common interface but do not have a common ancestor and require the same method implementation.
+A default implementation can be included in the interface to promote code reuse. Similarly, a certain class may not require certain methods listed in the interface.
+An example is a stub class that is required to implement an interface. A default method can be created with the `NotImplementedException`.
```
The explanation is too abstract. If you don't want to explain it in more detail (similar to the two features explained above), you can list this one under `Other features` and give a link to an external page that has a better explanation.
The same applies for Async/await.

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,246 @@
+
+Below are some common reasons why one should learn C#.
+
+### Cross platform support
```
`### Reason 1: Cross-platform support`

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,246 @@
+As Mono supports many platforms such as Windows, MacOS, Linux and even PlayStation 4, users can build for many platforms.
+C# is also used by the Unity Game Engine, which has high cross-platform support for game developers.
+
+### Popular Language
```
`### Reason 2: High in Demand`

damithc replied:
Add your explanatory sentence before citing others.

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,246 @@
+>Today, it is the 4th most popular programming language, with approximately 31% of all developers using it regularly. It is also the 3rd largest community on StackOverflow (which was built using C#) with more than 1.1 million topics.
+
+>[_Why Is C# Among The Most Popular Programming Languages in The World?_ - Armina Mkhitaryan](https://medium.com/sololearn/why-is-c-among-the-most-popular-programming-languages-in-the-world-ccf26824ffcb)
+
```
You can use `<blockquote>` tag to format a section as a quote.

damithc replied:
Format the quote in this way:
![image](https://user-images.githubusercontent.com/1673303/56079329-7fa79200-5e25-11e9-82dc-d314201269f6.png)

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,246 @@
+
+>[_It pays to learn to code with C# and here's why_ - Team Commerce](https://mashable.com/2018/03/17/coding-course-class-bootcamp/#om2xRzXFHGqJ)
+
+### Game Development
```
I would shorten this section section as follows:

>## Why Learn C#
>
>It is high in demand (as per [source1](), [source2]()). Especially well-suited for Windows apps and cross-platform games (via Unity).

Well, it can be a bit longer than that but shouldn't take more than a few sentences. The point about cross-platform nature can be omitted as it was already mentioned at the start.

damithc commented on contents/csharp/IntroductionToCSharp.md:
```diff
@@ -0,0 +1,246 @@
+realisation. You may reduce these problems by reading about [some common mistakes in C# programming](https://www.upwork.com/hiring/development/common-mistakes-in-c-sharp-programming/).
+
+If you want to develop a desktop application for Windows, you may consult [Microsoft's documentation](https://docs.microsoft.com/en-us/visualstudio/get-started/csharp/tutorial-wpf?view=vs-2019)
+on creating an application with the `Windows Presentation Foundation`, a framework that is commonly used for creating UI for Windows Applications that has many useful features for building your UI.
```
This section looks good. Perhaps drop the mention of Address Book? That's not supposed to be a big software anyway.

damithc reviewed (changes requested):
Getting better. Added some suggestions.

YuPeiHenry committed:
Minor reformatting.

YuPeiHenry committed:
Do bigger reformat edits.

YuPeiHenry committed:
Add explanation and content modifications.

YuPeiHenry commented:
@damithc Can you help me to do another round of review? I have addressed the issues that were previously raised.

damithc committed:
cosmetic tweaks

damithc committed:
fix typos

damithc commented:
Merged and deployed.
Good work @YuPeiHenry 
Thanks to reviewers too.

PS: I did some cosmetic tweaks and also shortened the WHY section.
